### DISCUSSION TITLE
What kind of an OptionROM image is generated if PCI Ven/Dev pair is not passed to the build command?

### DISCUSSION THREAD

**Discussion by Kostr** [ipxe/ipxe#404] [Q&A] [answered]
Hello!

It is possible to build an OptionROM IPXE image for the particular PCI device by providing a Vendor/Device pair to the build command, for example:
```
make bin/808610de.rom
make bin-x86_64-efi/808610de.efirom
```
But can someone say to me, what is happening if PCI Vendor/Device pair is not supplied to the build command?
```
make bin/ipxe.rom
make bin-x86_64-efi/ipxe.efirom
```
I couldn't build the `ipxe.rom` image, but the build for the `ipxe.efirom` has actually succeeded. Result of this build is a EFI PCI ROM image with a vendor/device pair set to `FFF3:0000`. What does it mean? And how this OptionROM image is supposed to work if it has vendor/device pair that doesn't match pci vendor/device pair?
Because usually BIOS code checks for the match between those two. Look at coreboot and seaBIOS source codes for example:
- Coreboot: https://github.com/coreboot/coreboot/blob/0a7193458556b556a4905a3551deb5003a57c9f9/src/device/pci_rom.c#L191
- Seabios: https://github.com/coreboot/seabios/blob/54082c81d96028ba8c76fbe6784085cf1df76b20/src/optionroms.c#L290

**Comment by NiKiZe** [comment]
Why do you ask?
Any it seems that failing to build ipxe.rom is a good thing.

Should iPXE have to add additional checks to make sure it is not built or used in a specific way?

**Reply by Kostr** [reply]
No particular reason, I guess I'm just trying to understand how does IPXE and PCI OptionROMs work.

Is this IPXE OptionROM build without Ven/Dev valid or not? 
- If it is valid, how it is supposed to work, if the produced image can't pass Ven/Dev check in BIOS?
- If it is not valid, shouldn't it be explicitly forbidden to make such images?

**Reply by NiKiZe** [reply]
You have looked at one bios implementation. Not all correct? And what about putting the .rom on a non PCI device and loaded anyway?

Why should iPXE impose artificial limitations?
The question is still, why are you trying to build this in the first place? Have anyone told that this _should_ work?
We have other options that don't make sense as well, undionly.lkrn, and all the .kkpxe and .kkkpxe should almost never be used.

**Reply by Kostr** [reply]
I've looked at edk2 also, it is all the same) 
But I understand your point. I've never thought that this image can be thought as a rom for a non PCI device. Especially if it has PCI rom signatures.

Overall I guess I got confused from this doc string: (https://ipxe.org/appnote/buildtargets)
```
ipxe builds one binary with (almost) all PCI based NIC drivers that iPXE has (for now not working on arm since some interfaces have not yet been implemented there, build specific drivers instead such as snp)
```
So I initially thought that something like
```
make bin/ipxe.rom
make bin-x86_64-efi/ipxe.efirom
```
should build me the rom image that would work as a PCI image for any PCI device.

**Comment by habetsm-xilinx** [comment]
When I try to build ipxe.rom it simply gets too big.
Build ipxe.mrom in stead. mrom images get around the size limit that rom images have.

To see what is in the resulting image attach ".info" to the build target, e.g.
make bin/ipxe.mrom.info.
You'll see it does not have a specific vendor/device pair, since it can bind to many devices.

**Reply by NiKiZe** [reply]
"it simply get too big" or rather "your flash space is to small for all the features you have chosen to enable" 😉

**Reply by habetsm-xilinx** [reply]
It is not related to the flash space.
Cannot quite remember, but I think there is something in the BIOS where it maps only the first 64KB of a PCI Expansion ROM. An mrom image can be used to get around this, it acts as as a small pre-loader.

**Comment by mcb30** [comment, marked as answer]
>     * If it is not valid, shouldn't it be explicitly forbidden to make such images?

In common with Unix: iPXE was not designed to stop you from doing stupid things, because that would also stop you from doing clever things.

**Reply by Kostr** [reply]
Thanks, that is a good point!

Now I just wonder if my stupid case should be marked with a less random Ven/Dev pair than `FFF3:0000`. Woudn't something like `0000:0000` or `ffff:ffff` be better in telling that there is no actual Ven/Dev pair in this PCI Oprom image?

**Reply by mcb30** [reply]
I'm not sure where your `fff3:0000` is coming from.  If I try `make bin/ipxe.mrom` and then dump the information using `./util/disrom.pl bin/ipxe.mrom`, it shows:
```
$ make bin/ipxe.mrom && ./util/disrom.pl bin/ipxe.mrom 
make: 'bin/ipxe.mrom' is up to date.
Index: 0, offset: 0x00000000

ROM header:

  Length:          0x08 (4096)
  Checksum:        0x0a (0x00)
  Init:            0x00a8
  UNDI header:     0x0084
  PCI header:      0x001c
  PnP header:      0x0040

PCI header:

  Signature:       PCIR
  Vendor ID:       0x0000
  Device ID:       0x0000
  Device list:     
  Device class:    0x020000
  Image length:    0x0008 (4096)
  Runtime length:  0x0008 (4096)
  Code type:       0x00
  Config header:   0x0000
  CLP entry:       0x0000
```
i.e. the vendor/device ID is (as expected) coming out as `0000:0000`.

There is no PCI device in the upstream iPXE codebase that uses vendor `fff3`.  Are you using a modified codebase?

**Reply by Kostr** [reply]
@mcb30 Hi!
As I said in my original message I've spotted this Ven/Dev pair in the `ipxe.efirom` image that is generated from this command:
```
make bin-x86_64-efi/ipxe.efirom
````
I've checked it once again with the master code from git, the issue is still present:
```
$ hexdump -C -n 64 bin-x86_64-efi/ipxe.efirom
00000000  55 aa d9 03 f1 0e 00 00  0b 00 64 86 01 00 00 00  |U.........d.....|
00000010  00 00 00 00 00 00 38 00  1c 00 00 00 50 43 49 52  |......8.....PCIR|
00000020  f3 ff 00 00 00 00 18 00  00 00 00 02 d9 03 00 00  |................|
00000030  03 80 00 00 60 00 00 00  75 b0 07 00 00 5c 0f 00  |....`...u....\..|
0000004
```

**Reply by mcb30** [reply]
> @mcb30 Hi!
> As I said in my original message I've spotted this Ven/Dev pair in the `ipxe.efirom` image that is generated from this command:
> 
> ```
> make bin-x86_64-efi/ipxe.efirom
> ```

Ah, sorry, I missed that.  I can indeed reproduce the `fff3` for `bin-x86_64-efi/ipxe.efirom`.  It turns out to arise from the vendor and device IDs getting expanded to empty strings, resulting in the command:
```
./util/efirom -v  -d  -c bin-x86_64-efi/ipxe.efidrv bin-x86_64-efi/ipxe.efirom
```
This then gets interpreted as a vendor ID of `-d`, i.e. `-0xd == 0xfff3`.

**Reply by Kostr** [reply]
That is a fun bug) Good catch!

I can't mark your comment as an answer, but now my issue is fully resolved.

Thanks for the help!

**Reply by mcb30** [reply]
@Kostr You're welcome!  I've pushed a fix for this as https://github.com/ipxe/ipxe/commit/db6310c3e
